_posts/2019-01-26-generic-way-of-handling-loading-saving-and-validationerrors-in-angular.md

```diff
@@ -0,0 +1,410 @@
+    this.userService.update(user).subscribe(res => {
+        // TODO: set acting to false
+        // TODO: handle validation errors
+    })
+}
+add(user): void {
+    // TODO: set acting to true
+    this.userService.add(user).subscribe(res => {
+        // TODO: set acting to false
+        // TODO: handle validation errors
+    })
+}
+```
+
+To handle validation errors we have to check if the httpstatusCode is 400, manually map the data etc.
+
+These codesamples are in this article to prove a point. **It's dirty redundant logic that we have to implement over and over again**
+
+## Let's clean this up
+
+To achieve this we will use an angular *service* in combination with an angular *interceptor* and typescript *decorators*.
+
+The first thing we need is a `HttpStatusService` that exposes 3 observables:
+
+- `loading$`: whether the user is fetching data
+- `acting$`: whether the user is removing, updating or adding data
+- `validationErrors$`: whether there are validation errors or not
+
+```typescript
+// http-status.service.ts
+@Injectable({
+  // important to provide this service to the 
+  // injector of the root module 
+  providedIn: 'root'
+})
+export class HttpStatusService {
+  // regular subject because we don't want to replay
+  // the validationerrors
+  private validationErrorsSub$ 
+    = new Subject<ValidationError[]>();
+
+  // 2 subjects that replays the last value 
+  // (ideal for state)
+  private loadingSub$ = new ReplaySubject<boolean>(1);
+  private actingSub$ = new ReplaySubject<boolean>(1);
+
+  // we don't want to expose the subject for
+  // encapsulation purposes. That's why we convert them
+  // into observables
+  validationErrors$ = 
+    this.validationErrorsSub$.asObservable();
+  loading$ = 
+    this.loadingSub$.pipe(distinctUntilChanged());
+  acting$ = 
+    this.actingSub$.pipe(distinctUntilChanged());
+
+  // these are just some regular setters to next 
+  // the values in our subjects
+  set validationErrors(errors: ValidationError[]) {
+    this.validationErrorsSub$.next(errors);
+  }
+
+  set loading(v: boolean) {
+    this.loadingSub$.next(v);
+  }
+
+  set acting(v: boolean) {
+    this.actingSub$.next(v);
+  }
+}
+```
+
+So we have a service that basically holds the state of our 3 statusses.
+Now we still have to make sure that the setters of these observables are being called at the right place and the right time.
+We don't want to manually implement that for every call, so let's create an interceptor for that.
+
+```typescript
+@Injectable({
+    providedIn: 'root'
+})
+export class HttpStatusInterceptor implements HttpInterceptor {
+  // keep track of the loading calls
+  private loadingCalls = 0; 
+  // keep track of the acting calls
+  private actingCalls = 0; 
+
+  constructor(
+    private httpStatusService: HttpStatusService
+  ) {}
+
+  private changeStatus(v: boolean, method: string): void {
+    if (['POST', 'PUT', 'DELETE', 'PATCH']
+      .indexOf(method) > -1) {
```

Comment:
`.contains(method)`

Reply:
Not supported by all browsers I guess... https://jsbin.com/jukuguvupi/1/edit?html,js,console